enhancements/network/in-cluster-dns-and-loadbalancers-on-more-platforms.md

```diff
@@ -0,0 +1,320 @@
+type Platform struct {
+    // PlatformName holds the arbitrary string representing the infrastructure
+    // provider name, expected to be set at the installation time. This field
+    // is solely for informational and reporting purposes and is not expected
+    // to be used for decision-making.
+    // +kubebuilder:default:="Unknown"
+    // +default="Unknown"
+    // +kubebuilder:validation:XValidation:rule="oldSelf == 'Unknown' || self == oldSelf",message="platform name cannot be changed once set"
+    // +optional
+    PlatformName string `json:"platformName,omitempty"`
+
+    // CloudControllerManager when set to external, this property will enable
+    // an external cloud provider.
+    // +kubebuilder:default:=""
+    // +default=""
+    // +kubebuilder:validation:Enum="";External
+    // +optional
+    CloudControllerManager CloudControllerManager `json:"cloudControllerManager,omitempty"`
+
+    // InClusterLoadBalancer is an optional feature that uses haproxy and
+    // keepalived as loadbalancers running in the cluster. Is is useful in
+    // environments where it is not possible or desirable to use loadbalancers outside
+    // of the cluster.
+    // +optional
+    InClusterLoadBalancer *InClusterLoadBalancer `json:"inClusterLoadBalancer,omitempty"`
+
+    // InternalDNS, when set, activates a DNS service running inside the cluster
+    // to provide DNS resolution internally. It is useful in environments where
+    // it is not possible or desirable to manage the cluster's internal DNS
+    // records in an external DNS system.
+    // +kubebuilder:default:=""
+    // +default=""
+    // +kubebuilder:validation:Enum="";CoreDNS
+    // +optional
+    InternalDNS InternalDNS `json:"internalDNS,omitempty"`
+}
+
+type InternalDNSType string
+
+const (
+    // CoreDNS is the default service used to implement internal DNS within a cluster.
+    CoreDNS InternalDNS = "CoreDNS"
+)
+```
+
+### Risks and Mitigations
+
+All of the components in question are already widely deployed in OpenShift
+clusters.
+
+### Drawbacks
+
+
+## Open Questions [optional]
+
+
+## Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+## Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+### Dev Preview -> Tech Preview
+
+
+### Tech Preview -> GA
+
+
+### Removing a deprecated feature
+
+
+## Upgrade / Downgrade Strategy
+
+No change to how the components are upgraded and/or downgraded today.
+
+## Version Skew Strategy
+
+No change.
+
+## Operational Aspects of API Extensions
+
+This proposal will enable clusters installed in the future to have fewer CRDs,
+since they'll be able to use in-cluster network services without having to
+select the `baremetal` platform type. Thus, the unused CRDs from the
+`baremetal` platform won't be present on those clusters.
+
+## Support Procedures
+
+No new support implications.
+
+## Alternatives
+
+### Move In-Cluster Network Settings Out of the Platform Spec
```

Comment:
FWIW, in a perfect world I'd like to see this happen, for a couple of reasons:

- There's a ton of duplication in the API already, and we're probably only going to add support for more platforms over time. Each platform requires us to duplicate the entire in-cluster network API again.
- We anticipate having more cross-platform configuration settings in the near future and it would be nice to have a logical place to put them already present.

That said, I recognize this could complicate things quite a lot, as discussed in the cons section. I guess you can consider this a vote in favor of a common config section, but not a vote against the proposal above.